operators/s3_to_redshift_operator.py

```diff
@@ -402,3 +413,98 @@ def create_if_not_exists(self, schema, pg_hook, temp=False):
                        sortkey=sk)
 
         pg_hook.run([create_schema_query, create_table_query])
+
+    def mysql_to_redshift_type_convert(self, mysql_type):
+        mysql_type = mysql_type.lower()
+        try:
+            sql_type, extra = mysql_type.strip().split(" ", 1)
+
+            # This must be a tricky enum
+            if ')' in extra:
+                sql_type, extra = mysql_type.strip().split(")")
+
+        except ValueError:
+            sql_type = mysql_type.strip()
+            extra = ""
+
+        # check if extra contains unsigned
+        unsigned = "unsigned" in extra
+        # remove unsigned now
+        extra = re.sub("character set [\w\d]+\s*", "", extra.replace("unsigned", ""))
+        extra = re.sub("collate [\w\d]+\s*", "", extra.replace("unsigned", ""))
+        extra = extra.replace("auto_increment", "")
+        extra = extra.replace("serial", "")
+        extra = extra.replace("zerofill", "")
+        extra = extra.replace("unsigned", "")
+
+        if sql_type == "tinyint(1)":
```

Comment:
Rather than a ton of `elif`, it would probably be cleaner to wrap this as an array of dictionaries.

Maybe something like:

```
conversion = [
{
 "mysql": "double",
 "redshift": "double precision"},
{
 "mysql": "smallint(",
 "redshift":{"signed": "integer", "unsigned": "smallint"}
}]
````

@tedmiston What do you think of this?

Reply:
@szeevs @benjamingregory I like that idea in Python generally speaking but for this particular lib, I don't love maintaining an up-to-date mapping for (potentially) N x N databases with type edge cases... sounds error prone.  I'd rather we delegate the type mapping to a library that already cleanly maps MySQL &rarr; Redshift types.  It'd be nice to get a PostgreSQL &rarr; Redshift mapping in here too if possible.

What about using the dialects module in SQLAlchemy to create the mapping? It was MySQL and PostgreSQL out of the box.

http://docs.sqlalchemy.org/en/latest/core/type_basics.html#vendor-specific-types

And there's an add-on dialect for Redshift.

https://github.com/sqlalchemy-redshift/sqlalchemy-redshift.

FlyData has docs on mapping MySQL, Postgres &rarr; Redshift types as well.

https://www.flydata.com/resources/flydata-sync/data-type-mapping-mysql/

Reply:
@tedmiston @benjamingregory I don't think it's possible to use SQLAlchemy to map types, it just allows you to use the dialect of a specific database, but it doesn't help you to map mysql.BIGINT to relevant Redshift type. Regarding a map, it may be possible, but you will have to store some sort of regex for each datatype, which will make it very complex. FlyData's documentation, is not very accurate unfortunately. I understand you want to outsource the conversion, but currently there is no one else to handle it.